Specs/AESArm.lean

```diff
@@ -110,21 +110,20 @@ def SubWord (w : BitVec WordSize) : BitVec WordSize :=
 
 protected def InitKey {Param : KBR} (i : Nat) (key : BitVec Param.key_len)
   (acc : KeySchedule) : KeySchedule :=
-  if h₀ : Param.Nk ≤ i then acc
-  else
-    have h₁ : i * 32 + 32 - 1 - i * 32 + 1 = WordSize := by
-      simp only [WordSize]; omega
-    let wd := BitVec.cast h₁ $ extractLsb (i * 32 + 32 - 1) (i * 32) key
+  match i with
+  | 0 => acc
+  | i' + 1 =>
+    let i := Param.Nk - i
+    let wd := extractLsb' (i * 32) 32 key
     let (x:KeySchedule) := [wd]
-    have _ : Param.Nk - (i + 1) < Param.Nk - i := by omega
-    AESArm.InitKey (Param := Param) (i + 1) key (acc ++ x)
-  termination_by (Param.Nk - i)
+    AESArm.InitKey (Param := Param) i' key (acc ++ x)
```

Comment:
I'm having a bit of trouble following what this code does exactly, but I assume you've checked the new version computes the same value as the old version?

Reply:
This function divide `key` (For example for AES-128, `key` is of length 128 bit) into a list of 32-bits (called seeded `KeySchedule`). The original code goes from beginning to the end of the list, and when rewriting to use structural recursion, it became a bit weird. This is because structural recursion goes from larger `i` to smaller `i`.

Check the new commit. I rewrote this function to go from end of the list to the beginning, which simplified the code. 

The same thing happens in `KeyExpansion_helper`, but that function is a recurrence function where later values in the list depends on previously generated values. So, I couldn't just change it to go from end to the beginning.